File: app/src/main/java/com/github/goldy1992/mp3player/service/library/content/ContentRetrievers.java
package com.github.goldy1992.mp3player.service.library.content;

import com.github.goldy1992.mp3player.commons.MediaItemType;
import com.github.goldy1992.mp3player.service.library.MediaItemTypeIds;
import com.github.goldy1992.mp3player.service.library.content.retriever.ContentRetriever;
import com.github.goldy1992.mp3player.service.library.content.retriever.FoldersRetriever;
import com.github.goldy1992.mp3player.service.library.content.retriever.RootRetriever;
import com.github.goldy1992.mp3player.service.library.content.retriever.SongsFromFolderRetriever;
import com.github.goldy1992.mp3player.service.library.content.retriever.SongsRetriever;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class ContentRetrievers {

    /** */
    Map<Class<? extends ContentRetriever>, ContentRetriever> contentRetrieverMap;
    /** */
    Map<String, ContentRetriever> idToContentRetrieverMap;

    private final RootRetriever rootRetriever;

    @Inject
    public ContentRetrievers(MediaItemTypeIds mediaItemTypeIds,
                             RootRetriever rootRetriever,
                             SongsRetriever songsRetriever,
                             FoldersRetriever foldersRetriever,
                             SongsFromFolderRetriever songsFromFolderRetriever) {

        Map<Class<? extends ContentRetriever>, ContentRetriever> mapToReturn = new HashMap<>();
        mapToReturn.put(RootRetriever.class, rootRetriever);
        mapToReturn.put(SongsRetriever.class, songsRetriever);
        mapToReturn.put(FoldersRetriever.class, foldersRetriever);
        mapToReturn.put(SongsFromFolderRetriever.class, songsFromFolderRetriever);
        this.contentRetrieverMap = mapToReturn;
                createIdMap(mediaItemTypeIds);
        this.rootRetriever = rootRetriever;
    }

    public ContentRetriever get(String id) {
        return this.idToContentRetrieverMap.get(id);
    }

    public ContentRetriever get(Class<? extends ContentRetriever> clazz) {
        return this.contentRetrieverMap.get(clazz);
    }

    private void createIdMap(MediaItemTypeIds mediaItemTypeIds) {
        this.idToContentRetrieverMap = new HashMap<>();
        for (MediaItemType mediaItemTypeInfo : MediaItemType.values()) {
            String key = mediaItemTypeIds.getId(mediaItemTypeInfo);
            if (null != key) {
                switch (mediaItemTypeInfo) {
                    case SONGS:
                    case SONG:
                        addToIdToContentRetrieverMap(key, SongsRetriever.class);
                        break;
                    case FOLDER:
                        addToIdToContentRetrieverMap(key, SongsFromFolderRetriever.class);
                        break;
                    case FOLDERS:
                        addToIdToContentRetrieverMap(key, FoldersRetriever.class);
                        break;
                    case ROOT:
                        addToIdToContentRetrieverMap(key, RootRetriever.class);
                        break;
                    default:
                        break;
                }
            }
        }
    }

    public RootRetriever getRoot() {
        return rootRetriever;
    }

    private void addToIdToContentRetrieverMap(String key, Class<? extends ContentRetriever> clazz) {
        ContentRetriever contentRetriever = contentRetrieverMap.get(clazz);
        if (null != contentRetriever) {
            idToContentRetrieverMap.put(key, contentRetriever);

        }
    }
}

File: service/src/main/java/com/github/goldy1992/mp3player/service/library/content/retriever/ContentRetrievers.kt
package com.github.goldy1992.mp3player.service.library.content.retriever

import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.service.library.MediaItemTypeIds
import dagger.hilt.android.scopes.ServiceScoped
import javax.inject.Inject

@ServiceScoped
class ContentRetrievers

    @Inject
    constructor(mediaItemTypeIds: MediaItemTypeIds,
                private val rootRetriever: RootRetriever,
                private val songsRetriever: SongsRetriever,
                private val foldersRetriever: FoldersRetriever,
                private val songsFromFolderRetriever: SongsFromFolderRetriever,
                private val albumsRetriever: AlbumsRetriever,
                private val songsFromAlbumRetriever: SongsFromAlbumRetriever,
                ) {
    /**  */
    var contentRetrieverMap: Map<Class<out ContentRetriever>, ContentRetriever>
    /**  */
    var idToContentRetrieverMap: MutableMap<String, ContentRetriever>? = null
    val root: RootRetriever
    operator fun get(id: String?): ContentRetriever? {
        return idToContentRetrieverMap!![id]
    }

    operator fun get(clazz: Class<out ContentRetriever?>?): ContentRetriever? {
        return contentRetrieverMap[clazz]
    }

    private fun createIdMap(mediaItemTypeIds: MediaItemTypeIds) {
        idToContentRetrieverMap = HashMap()
        for (mediaItemTypeInfo in MediaItemType.values()) {
            val key = mediaItemTypeIds.getId(mediaItemTypeInfo)
            if (MediaItemType.NONE.name != key) {
                when (mediaItemTypeInfo) {
                    MediaItemType.SONGS, MediaItemType.SONG -> addToIdToContentRetrieverMap(key, SongsRetriever::class.java)
                    MediaItemType.FOLDER -> addToIdToContentRetrieverMap(key, SongsFromFolderRetriever::class.java)
                    MediaItemType.FOLDERS -> addToIdToContentRetrieverMap(key, FoldersRetriever::class.java)
                    MediaItemType.ALBUMS -> addToIdToContentRetrieverMap(key, AlbumsRetriever::class.java)
                    MediaItemType.ALBUM -> addToIdToContentRetrieverMap(key, SongsFromAlbumRetriever::class.java)
                    MediaItemType.ROOT -> addToIdToContentRetrieverMap(key, RootRetriever::class.java)
                    else -> {
                    }
                }
            }
        }
    }

    fun getContentRetriever(mediaItemType: MediaItemType) : ContentRetriever? {
        return when (mediaItemType) {
            MediaItemType.ROOT -> root
            MediaItemType.SONGS -> songsRetriever
            MediaItemType.FOLDER -> songsFromFolderRetriever
            MediaItemType.FOLDERS -> foldersRetriever
            MediaItemType.ALBUMS -> albumsRetriever
            MediaItemType.ALBUM -> songsFromAlbumRetriever
            else -> null

        }
    }

    private fun addToIdToContentRetrieverMap(key: String, clazz: Class<out ContentRetriever>) {
        val contentRetriever = contentRetrieverMap[clazz]
        if (null != contentRetriever) {
            idToContentRetrieverMap!![key] = contentRetriever
        }
    }

    init {
        val mapToReturn: MutableMap<Class<out ContentRetriever>, ContentRetriever> = HashMap()
        mapToReturn[RootRetriever::class.java] = rootRetriever
        mapToReturn[SongsRetriever::class.java] = songsRetriever
        mapToReturn[FoldersRetriever::class.java] = foldersRetriever
        mapToReturn[AlbumsRetriever::class.java] = albumsRetriever
        mapToReturn[SongsFromFolderRetriever::class.java] = songsFromFolderRetriever
        contentRetrieverMap = mapToReturn
        createIdMap(mediaItemTypeIds)
        this.root = rootRetriever
    }
}